llvm/tools/llvm-profdata/llvm-profdata.cpp

```diff
@@ -118,18 +118,18 @@ cl::opt<std::string> ProfiledBinary(
     "profiled-binary", cl::init(""),
     cl::desc("Path to binary from which the profile was collected."),
     cl::sub(ShowSubcommand), cl::sub(MergeSubcommand));
-cl::opt<std::string> DebugInfoFilename(
-    "debug-info", cl::init(""),
+cl::list<std::string> DebugInfoFilenames(
+    "debug-info",
     cl::desc(
         "For show, read and extract profile metadata from debug info and show "
         "the functions it found. For merge, use the provided debug info to "
         "correlate the raw profile."),
     cl::sub(ShowSubcommand), cl::sub(MergeSubcommand));
-cl::opt<std::string>
-    BinaryFilename("binary-file", cl::init(""),
-                   cl::desc("For merge, use the provided unstripped bianry to "
-                            "correlate the raw profile."),
-                   cl::sub(MergeSubcommand));
+cl::list<std::string>
+    BinaryFilenames("binary-file",
+                    cl::desc("For merge, use the provided unstripped bianry to "
+                             "correlate the raw profile."),
+                    cl::sub(MergeSubcommand));
```

Comment:
> It looks like we are switching from passing a single correlation file to a list of correlation files. Is that so that we can pass all raw profiles into a single llvm-profdata merge command?

Yes, that's the purpose. 

> Presumably we should know which binaries produce which raw profiles.

This isn't always the case. When there are shared libraries producing raw profiles and `LLVM_PROFILE_FILE=default-%4m.profraw`, we will get bunch of `default-{hash}.profraw` produced by all shared libraries and main executable. There's no way to differentiate which raw profiles are generated by which binaries. @gulfemsavrun also has use case for this situation: https://discourse.llvm.org/t/rfc-add-binary-profile-correlation-to-not-load-profile-metadata-sections-into-memory-at-runtime/74565/8?u=zequanwu

Reply:
Another option would be to extend the pattern strings to support `%b` to expand to the binary id or the binary name. Do you think that would work?

Reply:
For convenience, here are the docs: https://clang.llvm.org/docs/SourceBasedCodeCoverage.html#running-the-instrumented-program

Reply:
> Another option would be to extend the pattern strings to support %b to expand to the binary id or the binary name. Do you think that would work?

If the binary is built with binary id embedded, the raw profiles will also have binary id embedded. So there's no need to make `%b` expand to the binary id. 
The point for accepting multiple correlation files is to avoid some customizing scripts which associate raw profiles with different binaries. 

The output indexed profile file will contain profile information for multiple binaries. I'm not sure if this will cause problems/inconvenience for processing. What's your thoughts on this @gulfemsavrun @petrhosek?

Reply:
> > Another option would be to extend the pattern strings to support %b to expand to the binary id or the binary name. Do you think that would work?
> 
> If the binary is built with binary id embedded, the raw profiles will also have binary id embedded. So there's no need to make `%b` expand to the binary id. The point for accepting multiple correlation files is to avoid some customizing scripts which associate raw profiles with different binaries.
> 
> The output indexed profile file will contain profile information for multiple binaries. I'm not sure if this will cause problems/inconvenience for processing. What's your thoughts on this @gulfemsavrun @petrhosek?

In my experience, automated profile generation systems already need customized scripts to gather raw profile data for merging. I found that it isn't hard to group raw profiles by the binaries they come from and merge profiles for each binary individually. I would prefer this over making the profile parsing logic more complicated, which is already difficult to understand.

Reply:
This basically adds a map with keys being BuildID and values being correlators (constructed from the correlate files), and uses BuildID to match raw profiles and correlate files. 

For now, I'll try to enable binary correlation on our side by using customized scripts to see if it's easier to achieve this functionality.

Reply:
Sounds good, let me know how you're experience goes 😃